test/e2e/network/service.go

```diff
@@ -3914,6 +3914,154 @@ var _ = common.SIGDescribe("Services", func() {
 		}
 	})
+
+	ginkgo.It("should implement NodePort and HealthCheckNodePort correctly when ExternalTrafficPolicy changes", func(ctx context.Context) {
+		// This uses a LoadBalancer service but only tests the parts of it that are
+		// managed by the service proxy; there is another test in loadbalancer.go
+		// to test the load balancer parts.
+
+		namespace := f.Namespace.Name
+		serviceName := "external-local-update"
+		jig := e2eservice.NewTestJig(cs, namespace, serviceName)
+
+		nodes, err := e2enode.GetBoundedReadySchedulableNodes(ctx, cs, 3)
+		framework.ExpectNoError(err, "listing nodes")
+		if len(nodes.Items) != 3 {
+			e2eskipper.Skipf("Need at least 3 schedulable nodes for this test")
+		}
+
+		ginkgo.By("creating the service")
+		svc, err := jig.CreateLoadBalancerServiceWaitForClusterIPOnly(func(svc *v1.Service) {
+			svc.Spec.ExternalTrafficPolicy = v1.ServiceExternalTrafficPolicyLocal
+		})
+		framework.ExpectNoError(err, "creating the service")
+		nodePortStr := fmt.Sprintf("%d", svc.Spec.Ports[0].NodePort)
+		hcNodePortStr := fmt.Sprintf("%d", svc.Spec.HealthCheckNodePort)
+		framework.Logf("NodePort is %s, HealthCheckNodePort is %s", nodePortStr, hcNodePortStr)
+
+		ginkgo.By("creating an endpoint for the service")
+		_, err = jig.Run(ctx, nil)
+		framework.ExpectNoError(err, "creating the endpoint")
+
+		var endpointNode, endpointNodeIP string
+		endpointsNodeMap, err := getEndpointNodesWithInternalIP(ctx, jig)
+		framework.ExpectNoError(err, "fetching endpoint node IP")
+		for node, nodeIP := range endpointsNodeMap {
+			framework.Logf("endpoint is on node %s (%s)", node, nodeIP)
+			endpointNode = node
+			endpointNodeIP = nodeIP
+			break
+		}
+
+		ginkgo.By("creating a HostNetwork exec pod on a different node from the endpoint")
+		execPod := launchHostExecPod(ctx, cs, namespace, "hostexec", &endpointNode)
+		hostExecPodNodeIP := execPod.Status.HostIP
+
+		ginkgo.By("finding a third node with neither the endpoint nor the hostexec pod")
+		var thirdNodeIP string
+		for i := range nodes.Items {
+			node := &nodes.Items[i]
+			ips := e2enode.GetAddresses(node, v1.NodeInternalIP)
+			if len(ips) < 1 {
+				framework.Failf("No internal ip found for node %s", node.Name)
+			}
+			if ips[0] != endpointNodeIP && ips[0] != hostExecPodNodeIP {
+				thirdNodeIP = ips[0]
+				break
+			}
+		}
+		gomega.Expect(thirdNodeIP).NotTo(gomega.Equal(""), "cluster has at least 3 nodes but could not find a third node IP?")
+
+		checkOneNodePort := func(nodeIP, nodePortStr string, expectResponding bool, expectResponse string, deadline time.Time) {
+			var responseStr string
+			if expectResponding {
+				if expectResponse == "" {
+					responseStr = "responds"
+				} else {
+					responseStr = fmt.Sprintf("responds with %q", expectResponse)
+				}
+			} else {
+				responseStr = "does not respond"
+			}
+
+			cmd := fmt.Sprintf("curl -g -i -s --connect-timeout 3 http://%s/clientip", net.JoinHostPort(nodeIP, nodePortStr))
```

Comment:
you ask for the clientip here to be returned but the checks bellow sees to match on the http status code

Reply:
ok, curl magic
```suggestion
                        // curl options
                        // -g, --globoff                                     Disable URL globbing with {} and []
                        // -i, --include                                     Include response headers in output
                        // -s, --silent                                      Silent mode
			cmd := fmt.Sprintf("curl -g -i -s --connect-timeout 3 http://%s/clientip", net.JoinHostPort(nodeIP, nodePortStr))
```

Reply:
oh, clientip is leftover from an earlier verson...
hm... though client IP should be reliably preserved for all three checks in the Local case (since it's hostExec pod to node IP) so I guess we can check it.

Reply:
We use `-s` pretty much everywhere, and `-g` everywhere we use IP addresses. But yes, the `-i` should be documented.

Reply:
clientip is ok, I just got mislead by that, sorry 

For the flags, the main concern was not the flags themselves, is just that I was not able to figure out how the function worked without knowning their values and reading below